extension/observer/ecsobserver/sd.go

```diff
@@ -17,41 +17,112 @@ package ecsobserver
 import (
 	"context"
 	"fmt"
+	"io/ioutil"
 	"time"
 
 	"go.uber.org/zap"
 )
 
-type ServiceDiscovery struct {
-	logger *zap.Logger
-	cfg    Config
+type serviceDiscovery struct {
+	logger   *zap.Logger
+	cfg      Config
+	fetcher  *taskFetcher
+	filter   *taskFilter
+	exporter *taskExporter
 }
 
-type ServiceDiscoveryOptions struct {
-	Logger *zap.Logger
+type serviceDiscoveryOptions struct {
+	Logger          *zap.Logger
+	FetcherOverride *taskFetcher // for test
 }
 
-func NewDiscovery(cfg Config, opts ServiceDiscoveryOptions) (*ServiceDiscovery, error) {
-	// NOTE: there are other init logic, currently removed to reduce pr size
-	return &ServiceDiscovery{
-		logger: opts.Logger,
-		cfg:    cfg,
+func newDiscovery(cfg Config, opts serviceDiscoveryOptions) (*serviceDiscovery, error) {
+	svcNameFilter, err := serviceConfigsToFilter(cfg.Services)
+	if err != nil {
+		return nil, fmt.Errorf("init serivce name filter failed: %w", err)
+	}
+	var fetcher *taskFetcher
+	if opts.FetcherOverride != nil {
+		fetcher = opts.FetcherOverride
+	} else {
+		fetcher, err = newTaskFetcher(taskFetcherOptions{
+			Logger:            opts.Logger,
+			Region:            cfg.ClusterRegion,
+			Cluster:           cfg.ClusterName,
+			serviceNameFilter: svcNameFilter,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("init fetcher failed: %w", err)
+		}
+	}
```

Comment:
This should probably be done by the caller and the `fetcher` provided directly here.  Again, this smells of incomplete dependency injection attempting to support tests, but not fully considered.

Reply:
changed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3785/commits/392a893222c9e17750af80680a3f178447c5ab81